### PR TITLE
Geoid2

### PR DESCRIPTION
**Pull request by qdnguyen**
Description

This PR is needed to visualize geoid color and DTM.

The changes were done 
- Add zfactor to deform geoid DTM
- Add geoid example

Screenshots 

image

### CODE REVIEW

File: src/Renderer/Shader/TileVS.glsl [hunk 1]
```diff
@@ -14,6 +14,8 @@ uniform sampler2D   dTextures_00[1];
 uniform vec3        offsetScale_L00[1];
 uniform int         loadedTexturesCount[8];
 
+uniform float       zFactor;
```
Comment (peppsac) [inline src/Renderer/Shader/TileVS.glsl:17]: Please move all the `zFactor` stuff to a specific commit.

File: src/Renderer/Shader/TileVS.glsl [hunk 2]
```diff
@@ -59,7 +59,8 @@ void main() {
                 }
 
             #elif defined(DATA_TEXTURE_ELEVATION)
-                float   dv  = max(texture2D( dTextures_00[0], vVv ).w, 0.);
+                float   dv  = texture2D( dTextures_00[0], vVv ).w;
+                dv *=zFactor;
```
Comment (peppsac) [inline src/Renderer/Shader/TileVS.glsl]: You should be doing `dv *=zFactor;` below (just before `vPosition = vec4( position +  vNormal  * dv ,1.0 )`).
This way it would work with all elevation data sources.

**Review by peppsac — CHANGES_REQUESTED** [review]
Thanks for the rebase :)

Please use proper wording for your commits (see http://codeinthehole.com/writing/a-useful-template-for-commit-messages/ for some help) and split them in separate topics (e.g: 1 for the zFactor feature and 1 for the new example)

File: src/Renderer/Shader/TileVS.glsl (hunk 2 above)
Comment (qdnguyen) [reply src/Renderer/Shader/TileVS.glsl]: i did it before like you said and it work but i dont know why it does not work now if i put it before vPosition = vec4( position + vNormal * dv ,1.0 )). I think three is a problem with vNormal

File: .gitignore [hunk 3]
```diff
@@ -6,3 +6,5 @@
 /node_modules/
 /dist/
 /lib/
+/nbproject/private/
```
Comment (peppsac) [inline .gitignore]: This doesn't belong to the itowns's `.gitignore` - gitignore is only populated with folders/files that are by-product or temp files of itowns.

File: examples/geoid.html [hunk 4]
```diff
@@ -0,0 +1,71 @@
+<!DOCTYPE html>
+<!--
```
Comment (peppsac) [inline examples/geoid.html]: Please remove this generated header.

File: examples/geoid.html [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+To change this license header, choose License Headers in Project Properties.
+To change this template file, choose Tools | Templates
+and open the template in the editor.
+-->
+<html>
+    <head>
+        <title>Globe</title>
```
Comment (peppsac) [inline examples/geoid.html]: And fix the title :)

File: examples/geoid.html [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            var viewerDiv = document.getElementById('viewerDiv');
+            var globeView = new itowns.GlobeView(viewerDiv, positionOnGlobe);
+            var menuGlobe = new GuiTools('menuDiv', globeView);
+            menuGlobe.view = globeView;
+
+            var promises = []
+            promises.push(itowns.Fetcher.json('layers/JSONLayers/GeoidTexture.json').then(function(result) { return globeView.addLayer(result); }));
+            //promises.push(itowns.Fetcher.json('layers/JSONLayers/Ortho.json').then(function(result) { return globeView.addLayer(result); }));
+            promises.push(itowns.Fetcher.json('layers/JSONLayers/GeoidMNT.json').then(function(result) { return globeView.addLayer(result); }));
+            //promises.push(itowns.Fetcher.json('layers/JSONLayers/IGN_MNT.json').then(function(result) { return globeView.addLayer(result); }));
```
Comment (peppsac) [inline examples/geoid.html]: And remove dead code in all this file.

File: nbproject/project.properties [hunk 7]
```diff
@@ -0,0 +1,3 @@
+files.encoding=UTF-8
```
Comment (peppsac) [inline nbproject/project.properties]: This `nbproject` folder must not be added to iTowns.

File: src/Renderer/Shader/TileVS.glsl (hunk 2 above)
Comment (peppsac) [reply src/Renderer/Shader/TileVS.glsl]: Hmm.. I don't know what is the underlying issue here, but you need to fix this so the `z factor` feature works independently of the elevation data source.

File: utils/debug/Main.js [hunk 8]
```diff
@@ -2,3 +2,4 @@ export { default as Debug } from './Debug';
 export { default as PointCloudDebug } from './PointCloudDebug';
 export { default as createTileDebugUI } from './TileDebug';
 export { default as create3dTilesDebugUI } from './3dTilesDebug';
+
```
Comment (peppsac) [inline utils/debug/Main.js]: Do not add empty lines please.

File: src/Renderer/Shader/TileVS.glsl [hunk 9]
```diff
@@ -69,6 +70,7 @@ void main() {
             #endif
 
             vPosition = vec4( position +  vNormal  * dv ,1.0 );
+
```
Comment (peppsac) [inline src/Renderer/Shader/TileVS.glsl]: Do not add empty lines please.

File: src/Core/TileMesh.js [hunk 10]
```diff
@@ -169,7 +169,11 @@ TileMesh.prototype.center = function center() {
 };
 
 TileMesh.prototype.OBB = function OBB() {
-    return this.geometry.OBB;
+    var zFactor = this.material.uniforms.zFactor.value;
+    var OBB = this.geometry.OBB;
+        OBB.box3D.min.z *= zFactor;
```
Comment (peppsac) [inline src/Core/TileMesh.js]: Indentation is wrong here.

File: src/Renderer/LayeredMaterial.js [hunk 11]
```diff
@@ -491,5 +493,9 @@ LayeredMaterial.prototype.setSelected = function setSelected(selected) {
     this.uniforms.selected.value = selected;
 };
 
+LayeredMaterial.prototype.setLayerZFactor = function setLayerZFactor(zFactor) {
```
Comment (peppsac) [inline src/Renderer/LayeredMaterial.js]: Is this method useful? It's just a one-line method and you don't use it https://github.com/iTowns/itowns/pull/511/files#diff-caf5227be4f9e2f283d82df32a678e4bL523 so I'd suggest removing it for now.

**Review by peppsac — COMMENTED** [review]
Quick review, focusing mostly on style issues.

File: .gitignore (hunk 3 above)
Comment (autra) [reply .gitignore]: (You can add your own ignore stuff in the .gitignore in your home folder)

File: src/Renderer/Shader/TileVS.glsl (hunk 2 above)
Comment (nguyenign) [reply src/Renderer/Shader/TileVS.glsl]: i set zFactor value by default to 1, so it will not impact on elevation data.

File: nbproject/project.properties (hunk 7 above)
Comment (nguyenign) [reply nbproject/project.properties]: thanks for your advice

File: src/Renderer/Shader/TileVS.glsl (hunk 2 above)
Comment (peppsac) [reply src/Renderer/Shader/TileVS.glsl]: My point is: we want this to work regardless of the data source used for elevation.
As an example, `zFactor` feature should work on planar.html demo.

File: src/Core/TileMesh.js (hunk 10 above)
Comment (nguyenign) [reply src/Core/TileMesh.js]: don't understant! We need it to adapt Bbox to geoid tiles'bbox, zFactor is set by default to 1 so it will not have an impact on  DTM

File: src/Renderer/LayeredMaterial.js (hunk 11 above)
Comment (nguyenign) [reply src/Renderer/LayeredMaterial.js]: yes, i used it in updateLayeredMaterialNodeElevation to deform geoid

File: src/Renderer/Shader/TileVS.glsl (hunk 9 above)
Comment (nguyenign) [reply src/Renderer/Shader/TileVS.glsl]: done

File: utils/debug/Main.js (hunk 8 above)
Comment (nguyenign) [reply utils/debug/Main.js]: done

File: src/Core/TileMesh.js (hunk 10 above)
Comment (peppsac) [reply src/Core/TileMesh.js]: The [code indentation](https://en.wikipedia.org/wiki/Indentation_style) is wrong.

Btw, the code is wrong too since you're going to modify `OBB.box3D.min.z` (and max) each time the `OBB()` function is called.
So for instance if `min.z = 1.0` and `zFactor = 3.0` initially, `min.z` will be: `3` then `9` then `27`...

File: src/Renderer/LayeredMaterial.js (hunk 11 above)
Comment (peppsac) [reply src/Renderer/LayeredMaterial.js]: I meant: it's a one-liner and you only use once.
So I'd prefer you to remove it (and replace it's usage in `updateLayeredMaterialNodeElevation` with `material.uniforms.zFactor.value = layer.zFactor;`)

File: src/Renderer/Shader/TileVS.glsl (hunk 2 above)
Comment (qdnguyen) [reply src/Renderer/Shader/TileVS.glsl]: ok, done

File: src/Renderer/LayeredMaterial.js (hunk 11 above)
Comment (qdnguyen) [reply src/Renderer/LayeredMaterial.js]: done

**Comment by peppsac** [comment]
Just tested:
- you should upload your data to https://github.com/iTowns/iTowns2-sample-data, so you'll be able to use a URL that works for everyone (see https://github.com/iTowns/itowns/blob/master/examples/pointcloud.html#L105 for an example)
- you have a merge commit in your PR: you need to do a rebase
- your commits do not respect the git messages light convention we're trying to adopt: see https://github.com/iTowns/itowns/issues/518

Thanks!

File: src/Core/TileMesh.js (hunk 10 above)
Comment (qdnguyen) [reply src/Core/TileMesh.js]: is there a posibility that in TileMesh i can have a notification when zFactor is changed ?

Comment (qdnguyen) [reply src/Core/TileMesh.js]: @peppsac is there a copy function of OBB? what do you think if i make a copy of OBB and modify it?

Comment (peppsac) [reply src/Core/TileMesh.js]: IMHO, I don't think you want to build a copy of the OBB.
You should be able to use the OBB.scale attribute to store the `zFactor` (ie: `scale = (1, 1, zFactor)`).

Comment (nguyenign) [reply src/Core/TileMesh.js]: Thank @peppsac for help, i tried it before but it did not work (i have holes as you can see in image), i think i need to modify OBB max min to adapt bbox for SSE.

image

Comment (peppsac) [reply src/Core/TileMesh.js]: My reasonning is the following: `OBB` is an `Object3D`. If you can't get `Object3D` properties (like scale) to work then it's a bug in `OBB`. 
And this bug needs to be fixed (instead of adding another function to `OBB`).

Comment (nguyenign) [reply src/Core/TileMesh.js]: No, there is no bug and i understand what you said but I tried the scale and it did not work, because the problem is Box3D in OBB. Do not update it, the calculation of SSE will be wrong and the results as you can see in the picture above.

Btw, i also do not think the code above is wrong, because in the case there is no geoid, zFactor is always equal to 1, there is no effect on MNT. In geoid case, when zFactor is greater than 1, i saw it runs just fine as you can see in picture below.

image

Comment (nguyenign) [reply src/Core/TileMesh.js]: please dont take into account my last comment, i forgot update OBB, it work now by using OBB.scale and OBB.update .

**Comment by nguyenign** [comment]
ok, pr is ready to merge

File: src/Core/Prefab/GlobeView.js [hunk 12]
```diff
@@ -107,7 +107,7 @@ export function createGlobeLayer(id, options) {
         }
     }
 
-    const wgs84TileLayer = new GeometryLayer(id, options.object3d || new THREE.Group());
+    const wgs84TileLayer = new GeometryLayer(id, options.object3d);
```
Comment (peppsac) [inline src/Core/Prefab/GlobeView.js]: Why did you changed this?

File: src/Core/Prefab/GlobeView.js [hunk 13]
```diff
@@ -194,6 +194,8 @@ function GlobeView(viewerDiv, coordCarto, options = {}) {
     // Setup View
     View.call(this, 'EPSG:4978', viewerDiv, options);
 
+    options.object3d = options.object3d || this.scene;
```
Comment (peppsac) [inline src/Core/Prefab/GlobeView.js]: Same question: why did you change this?

Comment (peppsac) [reply src/Core/Prefab/GlobeView.js]: I don't think you need to modify GlobeView, except for adding the `setZFactor` method.

File: src/Core/TileMesh.js [hunk 14]
```diff
@@ -169,7 +169,11 @@ TileMesh.prototype.center = function center() {
 };
 
 TileMesh.prototype.OBB = function OBB() {
-    return this.geometry.OBB;
+    var zFactor = this.material.uniforms.zFactor.value;
+    var OBB = this.geometry.OBB;
+        OBB.scale.z = zFactor;
```
Comment (peppsac) [inline src/Core/TileMesh.js:175]: Code indentation is still wrong (https://github.com/iTowns/itowns/pull/511#discussion_r145419514).
And please use `const` instead of `var`.

File: utils/debug/Main.js [hunk 15]
```diff
@@ -1,4 +1,4 @@
 export { default as Debug } from './Debug';
 export { default as PointCloudDebug } from './PointCloudDebug';
 export { default as createTileDebugUI } from './TileDebug';
-export { default as create3dTilesDebugUI } from './3dTilesDebug';
+export { default as create3dTilesDebugUI } from './3dTilesDebug';
```
Comment (peppsac) [inline utils/debug/Main.js:4]: You don't need to modify this file.

**Review by peppsac — CHANGES_REQUESTED** [review]
This PR is improving, thanks.

You still need to fix a few minor issues.
Then you can rebase you commits on master and:
- keep only 2 commits:
    * 1 adding the `zFactor` feature
    * 1 adding an example
- make sure your commit wording matches the template from #518

File: src/Core/Prefab/GlobeView.js (hunk 12 above)
Comment (nguyenign) [reply src/Core/Prefab/GlobeView.js]: it's changed automatically when i did rebase, i will change it back!

File: src/Core/TileMesh.js (hunk 14 above)
Comment (nguyenign) [reply src/Core/TileMesh.js:175]: what wrong? it work for me!

**Comment by nguyenign** [comment]
ok, it's ready to merge now :)

**Comment by nosy-b** [comment]
@qdnguyen Any news on that nice feature?

**Comment by nguyenign** [comment]
no, as i said it's ready to merge!

**Comment by nosy-b** [comment]
@qdnguyen Could you rebase?

**Comment by nguyenign** [comment]
rebased!

**Comment by nosy-b** [comment]
You have some lint problems
  _### 302:1  error  Trailing spaces not allowed                                   no-trailing-spaces
  303:5  error  Expected exception block, space or tab after '//' in comment  spaced-comment_

Plus, could you add the example.html in the example/index.html page with the right screenshot as it's a new feature?

**Comment by gchoqueux** [comment]
moved in #1174
